### PR TITLE
feat: new middleware `redirect_as2`

### PR DESCRIPTION
**Pull request by kwaa**
## Description

This is an updated version of the same middleware in Aoba.

### Check List

- [x] Have you read the
      [CODE OF CONDUCT](https://github.com/lumeland/lume/blob/master/CODE_OF_CONDUCT.md)
- [x] Have you read the document
      [CONTRIBUTING](https://github.com/lumeland/lume/blob/master/CONTRIBUTING.md)
  - [x] One pull request per feature. If you want to do more than one thing,
        send multiple pull request.
  - [ ] Write tests.
  - [x] Run deno `fmt` to fix the code format before commit.
  - [ ] Document any change in the `CHANGELOG.md`.

### CODE REVIEW

**Comment by kwaa** [comment]
@oscarotero For this middleware, what do you think about renaming it to `redirect_as2` (`redirectAS2`) and providing the following options?

```ts
export type Options = (url: URL) => URL | undefined
```

```ts
// manual
server.use(redirectAS2(url => new URL(url.href, 'https://hatsu.local/posts/')))
```

```ts
// hatsu preset
export const hatsu = (instance: string) => (url: URL) => new URL(url.href, `https://${instance}/posts/`)

import Server from 'lume/core/server.ts'
import expires, { hatsu } from 'lume/middlewares/redirect_as2.ts'
const server = new Server()
server.use(redirectAS2(hatsu('hatsu.local')))
```

This would also be compatible with Bridgy Fed (not sure of the effect since I'm not currently using it):

```ts
// bridgy-fed preset
export const bridgyFed = () => (url: URL) => new URL(url.href, 'https://fed.brid.gy/r/')

import Server from 'lume/core/server.ts'
import expires, { bridgyFed } from 'lume/middlewares/redirect_as2.ts'
const server = new Server()
server.use(redirectAS2(bridgyFed))
```

**Comment by oscarotero** [comment]
Yep, I like the idea!
BTW, what does `as2` mean? 😄 

For consistency with other middlewares, I'd use an options object as the first argument, instead a function.
Also, why don't simply pass the url, instead of a function? something like:

```js
server.use(redirectAS2({
  dest: new URL("https://hatsu.local/posts/"),
}));
```

If the logic for calculate the final destination is different for hatsu and bridgy, maybe it can be configured in the `mode` option (name may change). For example:

```js
server.use(redirectAS2({
  dest: new URL("https://hatsu.local/posts/"),
  mode: "hatsu"
}));
```

(or automatically detect it from the domain name).
What do you think?

**Comment by kwaa** [comment]
> Yep, I like the idea! BTW, what does `as2` mean? 😄

[Activity Streams 2.0](https://www.w3.org/TR/activitystreams-core/)

> Also, why don't simply pass the url, instead of a function?

Using functions allows users who need it to write custom logic, for example:

```ts
server.use(redirectAS2(url => {
  const { origin, pathname, search } = new URL(url)
  if (url.pathname.includes('/posts') return new URL(`${origin}${pathname}`, 'https://hatsu.local/posts/')
}))
```

**Comment by oscarotero** [comment]
> Using functions allows users who need it to write custom logic

Yes, I know. My only concern is about the value provided by this middleware. Internally, it only has an `if` to check if the request is an activity stream request. That's why I was thinking of something more straightforward.
Maybe something like? `url: URL | (url: URL) => URL` (allowing to provide a url or a function returning an url).

**Comment by kwaa** [comment]
> Maybe something like? `url: URL | (url: URL) => URL` (allowing to provide a url or a function returning an URL).

If the URL is provided directly, the options like `instance`, `mode`, etc., need to be included as well. I think this might be more cumbersome than using functions.

**Comment by oscarotero** [comment]
Okay, let's start with the function and we can see later if an url option makes sense.

**Comment by kwaa** [comment]
> Okay, let's start with the function and we can see later if an url option makes sense.

I have updated PR.

File: middlewares/redirect_as2.ts [hunk 1]
```diff
@@ -0,0 +1,68 @@
+import type { Middleware } from "../core/server.ts";
+
+export type Options = (url: URL) => Promise<URL> | URL | undefined | void;
```
Comment (oscarotero) [inline middlewares/redirect_as2.ts:3]: Options should be an object for consistency with all other Lume's plugins and middlewares. It allows to extend it with extra options in the future.

File: middlewares/redirect_as2.ts [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+export interface BridgyFedOptions extends CommonOptions {
+  /**
+   * BridgyFed instance.
+   * @default `fed.brid.gy`
+   */
+  instance?: URL["host"];
+}
+
+export const hatsu = (options: HatsuOptions): Options => (url: URL) => {
```
Comment (oscarotero) [inline middlewares/redirect_as2.ts:36]: Good idea to provide some functions to rewrite the URL.
In this case, I think we don't need an options object but only the instance URL.

I'm thinking of something like this:

```js
server.use(redirect_as2({
  redirect: hatsu("https://hatsu.local"),
}));
```

**Review by oscarotero — CHANGES_REQUESTED** [review]

File: middlewares/redirect_as2.ts (hunk 2 above)
Comment (kwaa) [reply middlewares/redirect_as2.ts:36]: > In this case, I think we don't need an options object but only the instance URL.

The `rewriteUrl` function also uses the `host` param from `CommonOptions`.

Comment (oscarotero) [reply middlewares/redirect_as2.ts]: Ah, I see. And what about this?

```js
server.use(redirect_as2({
  location: new URL("https://example.com"), // If it's not specified, use request.url
  redirect: hatsu("https://hatsu.local"),
}));
```

And the redirect functions should do something like this:

```js
export function hatsu(instanceURL) {
  return (requestURL, instanceURL, siteURL) {
    // code here
  }
}
```

My point is to move the common options to the middleware. `hatsu` (or `bridgy`) is only a way to generate the redirection url, it should have only the hatsu specific configuration.

Comment (kwaa) [reply middlewares/redirect_as2.ts]: > My point is to move the common options to the middleware. `hatsu` (or `bridgy`) is only a way to generate the redirection url, it should have only the hatsu specific configuration.

Okay, I've moved `host` option to `Options`.

**Review by oscarotero — APPROVED** [review]

**Comment by oscarotero** [comment]
Great.
Thanks for your patience in the back and forth 🙌

**Pull request merged**